ENH: #8325 Add ability to create tables using the gbq module.

parthea1 opened the pull request:
closes #8325 
closes #10652
- Added ability to automatically create a table using the gbq.to_gbq function if destination table does not exist
- Added `gbq.gbq_table_exists()` function to the gbq module 
- Added `gbq.create_gbq_table()` function to the gbq module

jreback commented on pandas/io/gbq.py:
```diff
@@ -436,3 +498,57 @@ def generate_bq_schema(df, default_type='STRING'):
 
     return {'fields': fields}
+
+def gbq_table_exists(table, project_id=None):
```
name this ``table_exists`` or compat with other pandas names

jreback commented on pandas/io/gbq.py:
```diff
@@ -436,3 +498,57 @@ def generate_bq_schema(df, default_type='STRING'):
+
+    return connector.table_exists(dataset_id, table_id)
+
+def create_gbq_table(table, schema, project_id=None):
```
name this ``create_table``

jreback commented on pandas/io/gbq.py:
```diff
@@ -436,3 +498,57 @@ def generate_bq_schema(df, default_type='STRING'):
+    """
+
+    if not project_id:
+        raise TypeError("Missing required parameter: project_id")
```
you don't need this, just make it a required parameter

jreback commented on pandas/io/gbq.py:
```diff
@@ -371,13 +426,16 @@ def to_gbq(dataframe, destination_table, project_id=None, chunksize=10000,
-    response which can be quite long depending on the size of the insert.
-    There are several important limitations of the Google streaming API
-    which are detailed at:
+    A destination table will be created if the specified table does not exist.
```
split this up into paragraphs / bullet points

jreback commented on pandas/io/tests/test_gbq.py:
```diff
@@ -353,6 +352,21 @@ def test_generate_bq_schema(self):
         self.assertEqual(schema, test_schema)
 
+    def test_create_bq_table(self):
+        if PROJECT_ID and not missing_bq():
```
this is already done in the setup, it is needed again?

jreback commented on pandas/io/gbq.py:
```diff
@@ -407,6 +465,10 @@ def to_gbq(dataframe, destination_table, project_id=None, chunksize=10000,
     connector = GbqConnector(project_id, reauth = reauth)
     dataset_id, table_id = destination_table.rsplit('.',1)
 
+    #if table does not exist, create new table
```
- what happens if an existing table has a different schema?
- sql has this keyword ``if_exists``

```
if_exists : {'fail', 'replace', 'append'}, default 'fail'
    - fail: If table exists, do nothing.
    - replace: If table exists, drop it, recreate it, and insert data.
    - append: If table exists, insert data. Create if does not exist.
```

 I think you need it here (can you support ``replace``?)

parthea1 commented:
- Renamed ``gbq_table_exists`` to ``table_exists``
- Renamed ``create_gbq_table`` to ``create_table``
- Made ``project_id`` a required parameter in ``create_table``
- Reformatted descriptor text in ``to_gbq`` function into bullet points
- Removed redundant checks in test_gbq.py. eg.  ``if PROJECT_ID and not missing_bq()``
- Modify ``gbq_read`` to support the ``destination_table`` argument, which allows users to redirect the query results directly to a BigQuery table instead of a pandas DataFrame
- Add ``if_exists argument to ``gbq_read`` and ``to_gbq`` to support appending, or replacing a table in GoogleBig Query based on the argument provided.
    - If destination table has a different schema and ``if_exists is set to ``append``, raise ``InvalidSchemaError``
    - If destination table has a different schema and ``if_exists is set to ``replace``, delete and re-create destination table, wait 120 seconds before inserting data  (https://code.google.com/p/google-bigquery/issues/detail?id=191)
- Cleaned up code that does not adhere to PEP8
- Fixed issues with  ``setUp`` and ``tearDown`` in test_gbq.py

Ready for review. All unit tests are passing.

jreback commented on doc/source/io.rst:
```diff
@@ -3884,28 +3885,54 @@ destination DataFrame as well as a preferred column order as follows:
+
+   data_frame = pd.read_gbq('SELECT * FROM test_dataset.test_table',
+                             index_col='index_column_name',
+                             col_order=['col1', 'col2', 'col3'], project_id = projectid,
```
this should certainly not be an argument to ``pd.read_gbq`` which by definition returns a ``DataFrame``. pls remove this. Not again having this functionaily but it should clearly be defined and not added on to something else.

parthea1 replied:
I will restore pd.read_gbq to its original functionality. I like the idea of creating a new function ``pd.to_gbq_large()`` , which is intended for larger queries. The reason for adding the word large is because the the BigQuery allowLargeResult parameter requires a destination table to be specified. Should I create a new branch/PR for this feature? Please confirm the following looks good. 

function name: ``pd.to_gbq_large()``
```Arguments
    ----------
    query : str
        SQL-Like Query to return data values
    project_id : str
        Google BigQuery Account project ID.
    reauth : boolean (default False)
        Force Google BigQuery to reauthenticate the user. This is useful
        if multiple accounts are used.
    destination_table :  string (optional)
        Name of table to be written, in the form 'dataset.tablename'.
        When provided, the results will be redirected to provided destination table.
    if_exists : {'fail', 'replace', 'append'}, default 'fail'
            - fail: If table exists, do nothing.
            - replace: If table exists, drop it, recreate it, and insert data.
            - append: If table exists, insert data. Create if does not exist.

     -------

jreback replied:
for sure create a new pr
but not sure why you actually need a new function; why wouldn't  this just be part of to_gbq?

parthea1 replied:
I have prepared 2 proposals in order to allow the user to choose to upload a Pandas Dataframe to a BigQuery table or run a SQL-Like query and have the results goto a BigQuery table. I would prefer proposal 1. I will create a new PR once the decision is made.

######Proposal 1
 
Modify ``pd.to_gbq()`` to support both ``dataframe`` and ``query`` as arguments.  Both will be optional, but at least 1 is required. In the new revision the ``dataframe`` parameter will be changed to optional so that users can either specify to upload either a pandas DataFrame or run SQL-Like Query and have the results sent directly to a BigQuery table.

function name: ``pd.to_gbq()``
``` Arguments
    ----------
   dataframe : DataFrame (optional)
        DataFrame to be written. Note: one of dataframe or query is required.
    query : str (optional)
        SQL-Like Query to return data values. Note: one of dataframe or query is required.
    destination_table : string 
        Name of table to be written, in the form 'dataset.tablename'
    project_id : str
        Google BigQuery Account project ID.
    chunksize : int (default 10000)
        Number of rows to be inserted in each chunk from the dataframe.
    verbose : boolean (default True)
        Show percentage complete
    reauth : boolean (default False)
        Force Google BigQuery to reauthenticate the user. This is useful
        if multiple accounts are used.
    if_exists : {'fail', 'replace', 'append'}, default 'fail'
            - fail: If table exists, do nothing.
            - replace: If table exists, drop it, recreate it, and insert data.
            - append: If table exists, insert data. Create if does not exist.
    """

    # only one will survive
    if dataframe and query:
        raise AssertionError('Only one of dataframe and query can be provided')

   # at least one of dataframe and query is required
      if not (dataframe or query):
         raise AssertionError('At least one of dataframe and query can be provided')
``` 


######Proposal 2
- ``pd.to_gbq()`` will have still ``dataframe`` as a required parameter (i.e. no change )
- create a new ``pd`` function to support running a SQL-Like Query and having the results go directly to a BigQuery destination table. This can be used for running large queries which require the BigQuery allowLargeResults parameter.
Proposed function name :   ``pd.to_gbq_large``

jreback commented on doc/source/io.rst:
```diff
@@ -3927,6 +3954,20 @@ will produce the dictionary representation of the schema.
+generate_bq_schema function.
+
+.. code-block:: python
+    gbq.create_table('my_dataset.my_table', schema, project_id = projectid)
```
you need a blank line after these directives

jreback commented on doc/source/io.rst:
```diff
@@ -3884,28 +3885,54 @@ destination DataFrame as well as a preferred column order as follows:
+
+By default, rows are streamed to BigQuery in chunks of 10,000 rows,
+but you can pass other chunk values via the ``chunksize`` argument.
+You can also see the progess of your post via the ``verbose`` flag
```
break this up to several paragraphs

jreback commented on doc/source/io.rst:
```diff
@@ -3884,28 +3885,54 @@ destination DataFrame as well as a preferred column order as follows:
+
+Finally, you can also send data to a BigQuery table from a pandas DataFrame using the :func:`~pandas.io.to_gbq`
+function.
+
```
this should be a separate sub-section

jreback commented on pandas/io/gbq.py:
```diff
@@ -79,14 +85,39 @@ class InvalidColumnOrder(PandasError, IOError):
 
+
+class TableAlreadyExists(PandasError, IOError):
+    """
```
these should all be sub-classes of ``ValueError``, and nothing else

jreback commented on pandas/io/gbq.py:
```diff
@@ -144,26 +175,55 @@ def run_query(self, query):
-                }
+        job_data = {'configuration': {'query': {}}}
+
+        if not destination_table:
```
see above

jreback commented on pandas/io/gbq.py:
```diff
@@ -230,40 +291,129 @@ def load_data(self, dataframe, dataset_id, table_id, chunksize, verbose):
             if (len(rows) % chunksize == 0) or (remaining_rows == 0):
                 if verbose:
-                    sys.stdout.write("\rStreaming Insert is {0}% Complete".format(((total_rows - remaining_rows) * 100) / total_rows))
+                    sys.stdout.write("\rStreaming Insert is {0}% Complete".format(
```
you can just use print

jreback commented on pandas/io/gbq.py:
```diff
@@ -230,40 +291,129 @@ def load_data(self, dataframe, dataset_id, table_id, chunksize, verbose):
                 if verbose:
-                    sys.stdout.write("\rStreaming Insert is {0}% Complete".format(((total_rows - remaining_rows) * 100) / total_rows))
+                    sys.stdout.write("\rStreaming Insert is {0}% Complete".format(
+                        ((total_rows - remaining_rows) * 100) / total_rows))
```
futher this printing should be controlled by the ``verbose`` paramater

parthea1 replied:
> further this printing should be controlled by the verbose paramater
I see that there is a check for ``verbose`` already  =>  ``if verbose: ``

jreback commented on pandas/io/gbq.py:
```diff
@@ -230,40 +291,129 @@ def load_data(self, dataframe, dataset_id, table_id, chunksize, verbose):
-                                                        datasetId = dataset_id,
-                                                        tableId = table_id,
-                                                        body = body).execute()
+                response = self.service.tabledata().insertAll(projectId=self.project_id, datasetId=dataset_id,
```
what you changed is much more readable. When you have lots of parameters then should be listed individually

jreback commented on pandas/io/gbq.py:
```diff
@@ -230,40 +291,129 @@ def load_data(self, dataframe, dataset_id, table_id, chunksize, verbose):
+        except HttpError as ex:
+            raise UnknownGBQException(json.loads(ex.content)['error'].get('errors', None))
+
+    def delete_table(self, dataset_id, table_id):
```
for example this *could* be exposed in the top-level as well.

jreback commented on pandas/util/testing.py:
```diff
@@ -1178,6 +1179,15 @@ def getMixedTypeDict():
 def makeMixedDataFrame():
     return DataFrame(getMixedTypeDict()[1])
 
+def makeMixedDataFramev2(test_size):
```
just put this in the gbq testing module then.

jreback commented:
@parthea1 did you mean to close this?

parthea1 commented:
No, I didn't mean to close it. I can't find the open button. I think it closed automatically when I reverted my last commit. I thought it would be easier to revert my last commit, and submit a fresh one that takes into account all the feedback. I'm working on the new commit now. I think the PR will open again once I add a commit. 

Regarding the allowLargeResults feature (i.e. running a google big query without returning a dataframe) , do you think its better to add the following comment into #10474 and continue the discussion there?

> I have prepared 2 proposals in order to allow the user to choose to upload a Pandas Dataframe to a BigQuery table or run a SQL-Like query and have the results goto a BigQuery table. I will create a new PR once the decision is made.

parthea committed:
ENH: Add ability to create tables using the gbq module. #8325

parthea commented:
My latest commit takes into account all the feedback in this PR, including: 
- Revert change to `read_gbq()`. (I will create a separate PR to support Google BigQuery allowLargeResult parameter)
- Update documentation in io.rst
- Add `delete_table()` function to the gbq module 
- Code refactoring

Ready for review. Ran unit tests in test_gbq. All tests passed.

parthea commented:
Let me know if there is any other work needed for this PR. I have a few new pull requests that I want to create after this one.
1. Remove dependency on bq command line tool in `test_gbq.py`
2. Add ability to create and delete datasets using `create_dataset()` and `delete_dataset()`
3. Add support for BigQuery allowLargeResults #10474

jreback commented on doc/source/io.rst:
```diff
@@ -3888,56 +3908,163 @@ destination DataFrame as well as a preferred column order as follows:
+
+
+Creating BigQuery Tables
+~~~~~~~~~~~~~~~~~~
```
need to underline the full width of the text

jreback commented on doc/source/io.rst:
```diff
@@ -3888,56 +3908,163 @@ destination DataFrame as well as a preferred column order as follows:
+    my_table_exists = pd.table_exists('my_dataset.my_table', project_id = projectid):
+
+.. note::
+    If you delete and re-create a BigQuery table with the same name, but different table schema,
```
blank line after a directive

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -218,7 +218,9 @@ Other enhancements
 - Support pickling of ``Period`` objects (:issue:`10439`)
 
-- ``DataFrame.apply`` will return a Series of dicts if the passed function returns a dict and ``reduce=True`` (:issue:`8735`).
+- Added ability to automatically create a table using the gbq.to_gbq function if destination table does not exist (:issue:`8325`).
```
use double-backticks around things like ``gbq.to_gbq``

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -218,7 +218,9 @@ Other enhancements
-- ``DataFrame.apply`` will return a Series of dicts if the passed function returns a dict and ``reduce=True`` (:issue:`8735`).
+- Added ability to automatically create a table using the gbq.to_gbq function if destination table does not exist (:issue:`8325`).
+- Added ability to replace an existing table and schema when calling the gbq.to_gbq function via the if_exists parameter (:issue:`8325`).
+- Added the following functions to the gbq module: ``gbq.table_exists()``, ``gbq.create_table()``  and ``gbq.delete_table()`` (:issue:`8325`).
```
add a mention in the breaking changes section that the exceptions are now ``ValueError`` rather than ``IOError``

jreback commented:
cc @jacobschaer can you have a look?

parthea commented:
I've updated the documentation with the required changes. Travis is green.

jreback commented on doc/source/io.rst:
```diff
@@ -3941,56 +3961,165 @@ destination DataFrame as well as a preferred column order as follows:
++-----+------------+-------+-------+-------+
+
+
+.. code-block:: python
```
to show s frame use an ipython block
use Timestamps ; show df.dtypes as well

jreback commented on doc/source/io.rst:
```diff
@@ -3941,56 +3961,165 @@ destination DataFrame as well as a preferred column order as follows:
+
+.. code-block:: python
+
+    pd.to_gbq(df, 'my_dataset.my_table', projectid)
```
don't we support df.to_gbq(..)

jreback commented on doc/source/io.rst:
```diff
@@ -3941,56 +3961,165 @@ destination DataFrame as well as a preferred column order as follows:
-As of 0.15.2, the gbq module has a function ``generate_bq_schema`` which
-will produce the dictionary representation of the schema.
+You can also see the progess of your post via the ``verbose`` flag which defaults to ``True``.
+For example:
```
progress

jreback commented on doc/source/io.rst:
```diff
@@ -3941,56 +3961,165 @@ destination DataFrame as well as a preferred column order as follows:
-   service.
+    The http response code of Google BigQuery can be successful (200) even if the append failed.
+    For this reason, if there is a failure to append to the table, the complete error response
+    from BigQuery is returned which can be quite long given it provides a status for each row.
```
why is this? u should then also check the response and raise an exception of this happened
silent failing is bad

parthea replied:
I haven't experienced the issue mentioned in the comment. It was an existing comment that I moved. I'm not sure the comment is still valid. Can I just remove the comment?

jreback replied:
sure - I would assume the spec would only return a success if it's actually successful? u less the API is more strict (and for example requires u to verify things)

jreback commented on doc/source/io.rst:
```diff
@@ -3941,56 +3961,165 @@ destination DataFrame as well as a preferred column order as follows:
+.. code-block:: python
+
+   df = pd.DataFrame({'A': [1.0]})
+   pd.generate_bq_schema(df, default_type='STRING')
```
use your example df from above (which has multiple dtypes) and show the output here (it's a code block so paste it in)

jreback commented on doc/source/io.rst:
```diff
@@ -3941,56 +3961,165 @@ destination DataFrame as well as a preferred column order as follows:
+''''''''''''''''''''''''
+
+As of 0.17.0, the gbq module has a function :func:`~pandas.io.create_table` which allows users
+to create a table in bigquery. The only requirement is that the dataset must already exsit.
```
shouldn't all of these fund references be pandas.io.gbq.*?

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -314,6 +314,10 @@ Other enhancements
 
     pd.concat([foo, bar, baz], 1)
 
+- Added ability to automatically create a table using the ``gbq.to_gbq()`` function if destination table does not exist (:issue:`8325`).
```
make a new sub section for all of the gbq changes. put in a pointer to the docs

jreback commented on pandas/core/frame.py:
```diff
@@ -810,19 +810,32 @@ def to_dict(self, orient='dict'):
+        # If the if_exists argument is set to replace, and the existing
+            table has a different schema, there is a delay of 2 minutes.
+            See Google BigQuery issue 191.
+            https://code.google.com/p/google-bigquery/issues/detail?id=191
```
make the links show by annotating (use back ticks and __)

jreback commented on pandas/core/frame.py:
```diff
@@ -839,13 +852,16 @@ def to_gbq(self, destination_table, project_id=None, chunksize=10000,
             Force Google BigQuery to reauthenticate the user. This is useful
             if multiple accounts are used.
-
+        if_exists : {'fail', 'replace', 'append'}, default 'fail'
```
versionsdded directive

jreback commented on pandas/io/gbq.py:
```diff
@@ -251,19 +281,110 @@ def load_data(self, dataframe, dataset_id, table_id, chunksize, verbose):
+
+    def table_exists(self, dataset_id, table_id):
+        try:
+            from apiclient.errors import HttpError
```
make a private function to check the imports

jreback commented on doc/source/io.rst:
```diff
@@ -3950,14 +3950,31 @@ The :mod:`pandas.io.gbq` module provides a wrapper for Google's BigQuery
-For specifics on the service itself, see `here <https://developers.google.com/bigquery/>`__
+.. warning::
+
+    To use this module, you will need a valid BigQuery account. Refer to the
```
this should line up with the w in warning

jreback commented on doc/source/io.rst:
```diff
@@ -3965,72 +3982,173 @@ function.
-         'integer_col_name' : [1],
-         'boolean_col_name' : [True]})
-   df.to_gbq('my_dataset.my_table', project_id = projectid)
+    df = pd.DataFrame({'my_string': list('abc'),
```
put the actual creation into an ``ipython``` block (e.g. the ``df = pd.DataFrame.....``)

then show

```
df
df.dtypes
```

put the ``df.to_gbq....`` only in the code block

jreback commented on doc/source/io.rst:
```diff
@@ -3965,72 +3982,173 @@ function.
+                      'my_float64': np.arange(4.0, 7.0),
+                      'my_bool1': [True, False, True],
+                      'my_bool2': [False, True, False],
+                      'my_dates': pd.date_range('now', periods=3).values})
```
don't use ``.values`` on the daterange (they would get inferred anyhow), just pass the index directly.

jreback commented on doc/source/io.rst:
```diff
@@ -3965,72 +3982,173 @@ function.
+
+.. code-block:: python
+
+    df = pd.DataFrame({'my_string': list('abc'),
```
you don't need this as you have it above (you can certainly put in a reference if you want to jump to it though)

jreback commented on doc/source/io.rst:
```diff
@@ -3965,72 +3982,173 @@ function.
+
+.. code-block:: python
+
+    gbq.delete_table('my_dataset.my_table', projectid)
```
complete aside, is their a way to 'list' the tables under the project id?

parthea replied:
There is a 'list' function in the BigQuery API which returns a list of tables under a specified dataset. The pandas.io gbq module doesn't currently support 'list'. This could be a useful feature for a future pandas release.
(https://cloud.google.com/bigquery/docs/reference/v2/tables/list)

jreback commented on doc/source/io.rst:
```diff
@@ -3965,72 +3982,173 @@ function.
+
+.. code-block:: python
+
+    my_table_exists = gbq.table_exists('my_dataset.my_table', project_id = projectid)
```
this is a code block, so dont't use ``my_table_exists``, just show the result (e.g. True). You can show this as a copy-paste session from an ipython prompt

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -40,6 +40,7 @@ Highlights include:
 - Development support for benchmarking with the `Air Speed Velocity library <https://github.com/spacetelescope/asv/>`_ (:issue:`8316`)
 - Support for reading SAS xport files, see :ref:`here <whatsnew_0170.enhancements.sas_xport>`
 - Removal of the automatic TimeSeries broadcasting, deprecated since 0.8.0, see :ref:`here <whatsnew_0170.prior_deprecations>`
+- Support the ability to create and delete tables in Google BigQuery, see :ref:`here <whatsnew_0170.gbq>`
```
back-ticks on ``Google BigQuery``

parthea commented:
My latest commit takes into account all the feedback in this PR, including:
- Changes to the documentation for the gbq module in `io.rst`
- Additional error handling for http error as well as other errors that can occur while streaming data to BigQuery
- Create a sub section in `doc/source/whatsnew/v0.17.0.txt` for enhancements to the `gbq` module
- Ensure all links are formatted correctly
- Add a separate function in the gbq module for checking imports
- Add unit test `test_upload_data_dataset_not_found()`

Ready for review. Ran unit tests in test_gbq. All tests passed.

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -251,6 +252,13 @@ has been changed to make this keyword unnecessary - the change is shown below.
+- Added ability to automatically create a table using the ``gbq.to_gbq()`` function if destination table does not exist (:issue:`8325`).
+- Added ability to replace an existing table and schema when calling the ``gbq.to_gbq()`` function via the ``if_exists`` argument (:issue:`8325`).
+- Added the following functions to the gbq module: ``gbq.table_exists()``, ``gbq.create_table()``  and ``gbq.delete_table()`` (:issue:`8325`).
 
```
put a pointer to the docs here as well (in io.rst)

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -835,6 +843,38 @@ Other API Changes
   raise ``ValueError``                All other public methods (names not beginning with underscores)
   ===============================     ===============================================================
 
+
```
move this to the BigQuery section as well

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -835,6 +843,38 @@ Other API Changes
+    ``TermsOfServiceNotAccepted``
+    ``UnknownGBQException``
+
+- The following exceptions in the gbq module have been renamed/added:
```
this separate list is not necessary, remove it. Only the 'custom' ones matter. You can put a list of these exceptions in a pointer in the doc, but don't list them.

jreback commented on pandas/io/gbq.py:
```diff
@@ -79,23 +87,158 @@ class InvalidColumnOrder(PandasError, IOError):
-
         except ImportError:
-            raise ImportError('Could not import Google API Client.')
+            raise ImportError('Missing required module')
```
do ``except ImportError as e:`` ,then raise something like ``ImportError("......gbq: {0}".format(str(e)))`` (e.g. you need to re-raise, but with a more informative message

jreback commented on pandas/io/gbq.py:
```diff
@@ -79,23 +87,158 @@ class InvalidColumnOrder(PandasError, IOError):
     """
     pass
 
+
```
why are you adding all of these exceptions? If an exception happens, just have it raise directly. No need to cover over the API. Only really useful exceptions should be done as custom.

jreback commented on pandas/io/gbq.py:
```diff
@@ -133,13 +267,72 @@ def get_service(self, credentials):
+        status = json.loads(ex.content)['error']
+        errors = status.get('errors', None)
+
+        if errors:
```
just raise a generic exception here (you can make 1 custom one if you want). and include the message/reason.

jreback commented on pandas/io/gbq.py:
```diff
@@ -133,13 +267,72 @@ def get_service(self, credentials):
+                if verbose:
+                    print(error_message)
+                else:
+                    raise StreamingInsertError(error_message + '\nEnable verbose logging to see all errors')
```
so this is a legitimate (and useful) custom exception.

jreback commented on pandas/io/gbq.py:
```diff
@@ -148,122 +341,201 @@ def run_query(self, query):
-                                           body=job_data).execute()
-            status = query_reply['status']
+            query_reply = job_collection.insert(projectId=self.project_id, body=job_data).execute()
         except AccessTokenRefreshError:
```
you can certainly *catch* exceptions raised within gbq, but usually best just to let them bubble up, unless you can actually handle them (e.g. if the credentials are screwed up, then nothing you can do, so don't even catch it in the first place). 

All that said if their is an exception (or family) of exceptions that could use a more intuitve error message then pls do so.

jreback commented:
@parthea1 let's scale back a lot on the exception handling code. Its not tested and doesn't add value over simply re-raising. 

As I said above catching a http error in the middle of an insertion (and raising a nice miessage is fine).

Just use ``ValueError`` for everything (a couple of custom exceptions are ok, e.g. AuthorizationError) or whatever, but not too many. Its just confusing and not tested.

overall looks good though.

If you can make those changes we can put this in 0.17.0

jreback commented:
@jorisvandenbossche can you have a look?

jorisvandenbossche commented on doc/source/io.rst:
```diff
@@ -3950,14 +3950,31 @@ The :mod:`pandas.io.gbq` module provides a wrapper for Google's BigQuery
+
+The key functions are:
+
+.. autosummary::
```
you have to add a `.. currentmodule:: pandas.io.gbq` before this autosummary (see how it is done in api.rst), and change it back to pandas afterwards

jorisvandenbossche commented on doc/source/io.rst:
```diff
@@ -3965,72 +3982,184 @@ function.
    projectid = "xxxxxxxx"
 
-   data_frame = pd.read_gbq('SELECT * FROM test_dataset.test_table', project_id = projectid)
+   data_frame = gbq.read_gbq('SELECT * FROM test_dataset.test_table', projectid)
```
if you change this to gbq, can you add the import as well?

jreback replied:
let's leave this as ``pd.`` its what we advertise as the reading function

jorisvandenbossche commented on doc/source/io.rst:
```diff
@@ -3965,72 +3982,184 @@ function.
+
+Assume we want to write a DataFrame ``df`` into a BigQuery table using :func:`~pandas.DataFrame.to_gbq`.
+
+.. ipython:: python
```
or you have to make this a static `code-block` like the others, or you have to remove the output (but I would go for the code-block, since all the others are that as well)

jorisvandenbossche replied:
or maybe better is to use `.. ipython::` instead of `.. ipython:: python`, then the code is not executed, but the formatting is more similar (`.. code-block:: python` does look a bit different)

jorisvandenbossche commented on doc/source/io.rst:
```diff
@@ -3965,72 +3982,184 @@ function.
+
+.. ipython:: python
+
+   df = pd.DataFrame({'my_string': list('abc'),
```
in case of using a code-block, a ``In [3]`` should be added here

jorisvandenbossche commented on doc/source/io.rst:
```diff
@@ -3965,72 +3982,184 @@ function.
+
+  In [4]: df
+  Out[4]: my_bool1	my_bool2	my_dates	my_float64	my_int64	my_string
+       0	True	False	2015-09-10 03:03:45.365258	4	1	a
```
the alignment is not fully correct I think

jreback replied:
i you are using an ipython block (it is ok here). Then don't need the ``[4]``, instead just ``df`` and it will be printed

jorisvandenbossche commented on doc/source/io.rst:
```diff
@@ -3965,72 +3982,184 @@ function.
-You can access the management console to determine project id's by:
-<https://code.google.com/apis/console/b/0/?noredirect>
+The ``if_exists`` argument can be used to dictate whether to ``fail``, ``replace``
+or ``append`` if the destination table already exists. The default value is ``fail``.
```
Can you quote the fail, replace, append? (so ```` ``'fail'`` ````, then it is more clear it is a value and not a keyword)

jorisvandenbossche commented on doc/source/io.rst:
```diff
@@ -3965,72 +3982,184 @@ function.
+
+For example:
+
+.. ipython:: python
```
same here: make the directive `.. ipython::` or `.. code-block::` (with `.. ipython:: python`, the code is executed, which is not possible here)

jorisvandenbossche commented on pandas/core/frame.py:
```diff
@@ -810,19 +810,32 @@ def to_dict(self, orient='dict'):
-        There are several important limitations of the Google streaming API
-        which are `here <https://developers.google.com/bigquery/streaming-data-into-bigquery>`__
+        Notes:
+        * The destination dataset id must already exist
```
Move this to a separate 'Notes' section?

jorisvandenbossche commented on pandas/core/frame.py:
```diff
@@ -810,19 +810,32 @@ def to_dict(self, orient='dict'):
+        * The destination dataset id must already exist
+        * If the table does not exist, a new table will be created.
+        * If the table exists, the if_exists parameter will dictate
+            whether to fail, replace or append
```
alignment (the second line should match the start of the sentence ('If')

jorisvandenbossche commented on pandas/core/frame.py:
```diff
@@ -839,13 +852,18 @@ def to_gbq(self, destination_table, project_id=None, chunksize=10000,
             Force Google BigQuery to reauthenticate the user. This is useful
             if multiple accounts are used.
+        if_exists : {'fail', 'replace', 'append'}, default 'fail'
+                - fail: If table exists, do nothing.
```
alignment, text should start at same level as the explanation for the other parameters (so 4 spaces behind 'if_exists')

jorisvandenbossche commented on pandas/io/gbq.py:
```diff
@@ -368,20 +493,32 @@ def read_gbq(query, project_id=None, index_col=None, col_order=None, reauth=Fals
-           verbose=True, reauth=False):
+
+def to_gbq(dataframe, destination_table, project_id, chunksize=10000, verbose=True, reauth=False, if_exists='fail'):
     """Write a DataFrame to a Google BigQuery table.
```
same remarks here as for the docstring in `frame.py`

jorisvandenbossche commented on pandas/io/gbq.py:
```diff
@@ -368,20 +493,32 @@ def read_gbq(query, project_id=None, index_col=None, col_order=None, reauth=Fals
-def to_gbq(dataframe, destination_table, project_id=None, chunksize=10000,
-           verbose=True, reauth=False):
+
+def to_gbq(dataframe, destination_table, project_id, chunksize=10000, verbose=True, reauth=False, if_exists='fail'):
```
PEP8 line length

jorisvandenbossche commented on pandas/io/gbq.py:
```diff
@@ -439,3 +596,70 @@ def generate_bq_schema(df, default_type='STRING'):
+
+def table_exists(table, project_id):
+    """ Check if a table exists in big query given a table and project_id
+
```
can you add ``.. versionadded:: 0.17.0`` here?

jorisvandenbossche replied:
same for create_table and delete_table

jorisvandenbossche commented:
@parthea1 Looks like a nice enhancement! I put a bunch of comments, but it are just small things on the documentation formatting

I agree with @jreback that all the different custom exception classes are not needed. Just a clear error message is way more important

parthea commented:
I've implemented all the changes. Waiting for Travis.

parthea commented:
Travis is green.

jreback commented on doc/source/io.rst:
```diff
@@ -3982,56 +4003,156 @@ destination DataFrame as well as a preferred column order as follows:
+.. code-block:: python
+
+   df.to_gbq('my_dataset.my_table', projectid, chunksize=10000, verbose=True)
+
```
add a sample session here (as you are in a code-block, just show the output of you doing this)

jreback commented on doc/source/io.rst:
```diff
@@ -3982,56 +4003,156 @@ destination DataFrame as well as a preferred column order as follows:
+
+.. code-block:: python
+
+   pd.create_table('my_dataset.my_table', schema, projectid)
```
should be ``gbq.create_table....``, this is not part of the ``pd`` namespace, ONLY ``pd.read_gbq/to_gbq``

jreback commented on doc/source/io.rst:
```diff
@@ -3982,56 +4003,156 @@ destination DataFrame as well as a preferred column order as follows:
+.. code-block:: python
+
+   In [10]: pd.generate_bq_schema(df, default_type='STRING')
+
```
gbq

jreback commented on doc/source/io.rst:
```diff
@@ -3982,56 +4003,156 @@ destination DataFrame as well as a preferred column order as follows:
+
+.. code-block:: python
+
+   pd.delete_table('my_dataset.my_table', projectid)
```
gbq

jreback commented on doc/source/io.rst:
```diff
@@ -3982,56 +4003,156 @@ destination DataFrame as well as a preferred column order as follows:
+
+.. code-block:: python
+
+   In [12]: pd.table_exists('my_dataset.my_table', projectid)
```
gbq

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -281,6 +281,15 @@ has been changed to make this keyword unnecessary - the change is shown below.
+
+Google BigQuery Enhancements
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+- Added ability to automatically create a table using the ``gbq.to_gbq()`` function if destination table does not exist. (:issue:`8325`).
```
make these all function links, e.g. `` :func:`pandas.io.gbq.to_gbq` ``

jreback commented on pandas/io/gbq.py:
```diff
@@ -148,122 +192,201 @@ def run_query(self, query):
-                            jobId=job_reference['jobId']).execute()
+        while not query_reply.get('jobComplete', False):
+            if verbose:
+                print('Waiting for job to complete...')
```
later enhancement I would print the time on the message as well

jreback replied:
much futher enhancement (future PR), ``verbose`` could be a callable, which you would execute with the message (rather than actual print statements)

parthea replied:
Sounds good. I have a few PRs for the release after 0.17.0

jreback commented:
@parthea1 couple of minor doc changes. if you could asap would be gr8. pls let me know when you push.

jreback commented:
https://travis-ci.org/pydata/pandas/jobs/79827353

look at the skipped tests (one of the tabs at the very bottom). all of the gbq tests are being skipped because of an import issue.

can you pls fix this as well

jreback commented:
but test at least import for me (e.g. ``Successfully installed google-api-python-client-1.4.1 httplib2-0.9.1 oauth2client-1.5.0 pyasn1-0.1.8 pyasn1-modules-0.0.7 rsa-3.2 simplejson-3.8.0 uritemplate-0.6
``)

```
test_should_be_able_to_get_a_bigquery_service (pandas.io.tests.test_gbq.TestGBQConnectorIntegration) ... SKIP: Cannot run integration tests without a project id
test_should_be_able_to_get_results_from_query (pandas.io.tests.test_gbq.TestGBQConnectorIntegration) ... SKIP: Cannot run integration tests without a project id
test_should_be_able_to_get_schema_from_query (pandas.io.tests.test_gbq.TestGBQConnectorIntegration) ... SKIP: Cannot run integration tests without a project id
test_should_be_able_to_get_valid_credentials (pandas.io.tests.test_gbq.TestGBQConnectorIntegration) ... SKIP: Cannot run integration tests without a project id
test_should_be_able_to_make_a_connector (pandas.io.tests.test_gbq.TestGBQConnectorIntegration) ... SKIP: Cannot run integration tests without a project id
SKIP: Cannot run integration tests without a project id
test_read_gbq_with_no_project_id_given_should_fail (pandas.io.tests.test_gbq.TestReadGBQUnitTests) ... ok
test_should_return_bigquery_booleans_as_python_booleans (pandas.io.tests.test_gbq.TestReadGBQUnitTests) ... ok
test_should_return_bigquery_floats_as_python_floats (pandas.io.tests.test_gbq.TestReadGBQUnitTests) ... ok
test_should_return_bigquery_integers_as_python_floats (pandas.io.tests.test_gbq.TestReadGBQUnitTests) ... ok
test_should_return_bigquery_strings_as_python_strings (pandas.io.tests.test_gbq.TestReadGBQUnitTests) ... ok
test_should_return_bigquery_timestamps_as_numpy_datetime (pandas.io.tests.test_gbq.TestReadGBQUnitTests) ... ok
test_that_parse_data_works_properly (pandas.io.tests.test_gbq.TestReadGBQUnitTests) ... ok
test_to_gbq_should_fail_if_invalid_table_name_passed (pandas.io.tests.test_gbq.TestReadGBQUnitTests) ... ok
test_to_gbq_with_no_project_id_given_should_fail (pandas.io.tests.test_gbq.TestReadGBQUnitTests) ... ok
SKIP: Cannot run integration tests without a project id
pandas.io.tests.test_gbq.test_requirements ... ok
```

parthea commented:
I've updated the docs with the minor changes.

I have only seen the following message when testing locally. After I enter the project Id, all tests passed.
"SKIP: Cannot run integration tests without a project id"

I haven't encountered the error "Could not import Google API Client" locally. I ran all units tests with a project Id and they all pass. 

I could look into it after 0.17.0 release

parthea commented:
I notice that the  `apiclient` is not installed 

``` "apiclient: None"```


```
INSTALLED VERSIONS
------------------
commit: 8174073eb24080bed4b3eb4b1d6f9283c3ba33d8
python: 2.7.10.final.0
python-bits: 64
OS: Linux
OS-release: 2.6.32-042stab108.8
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: en_US.UTF-8
LANG: en_US.UTF-8

pandas: 0.16.2+635.g8174073.dirty
nose: 1.3.7
pip: 7.1.2
setuptools: 18.1
Cython: 0.19.1
numpy: 1.7.0
scipy: 0.13.2
statsmodels: 0.5.0
IPython: 2.4.1
sphinx: 1.1.3
patsy: 0.2.1
dateutil: 1.5
pytz: 2013b
bottleneck: 0.8.0
tables: 3.0.0
numexpr: 2.2.2
matplotlib: 1.3.1
openpyxl: 1.6.2
xlrd: 0.9.2
xlwt: 0.7.5
xlsxwriter: 0.4.6
lxml: None
bs4: 4.2.1
html5lib: 1.0b2
httplib2: None
apiclient: None
sqlalchemy: 0.9.6
pymysql: 0.6.3.None
psycopg2: 2.5.2 (dt dec pq3 ext)
travis_time:end:1f1b75ec:start=1441971280601163542,finish=1441971282745902942,duration=2144739400
[0Ktravis_fold:end:after_script.3
[0Ktravis_fold:start:after_script.4
[0Ktravis_time:start:0330ff2b
[0K$ ci/print_skipped.py /tmp/nosetests.xml
```

jreback commented:
right I think we need to install it as 1.2.0

pls see what u can do

parthea commented:
Is it ok to use the latest version of google-api-python-client `1.4.2`, instead of `1.2`?

All unit tests in `test_gbq.py` pass in my local environment with `1.4.2`.

The only change I would recommend after updating to `1.4.2` is to change `apiclient` to `googleapiclient`. `apiclient` is an alias for `googleapiclient`

see https://github.com/google/google-api-python-client/blob/master/apiclient/__init__.py

jreback commented:
no see if u can make the original one work
not averse to adding the API stuff to snothrr build though

jreback commented:
I am going to merge this: https://github.com/pydata/pandas/pull/11081/files

shortly - you can just put the things (and exact versions) in the ``*.pip`` files

parthea commented:
Great! I'm hoping my most recent commit will resolve the `Could not import Google API Client` import error. 

While I was resolving the import error, I ran into the issue mentioned in #10656 where `LooseVersion` does not treat `'1.2'` and `'1.2.0'` as equal. I think the solution is to use `StrictVersion`. I've added the commit that fixes the version check.

Here is a comparison of `LooseVersion` and `StrictVersion`
```
>>> from distutils.version import LooseVersion
>>> a = LooseVersion('1.2')
>>> b = LooseVersion('1.2.0')
>>> a == b
False
>>> from distutils.version import StrictVersion
>>> c = StrictVersion('1.2')
>>> d = StrictVersion('1.2.0')
>>> c == d
True
```

jreback commented:
ok gr8!

having tests that work on Travis will be a big help 

even better would be a test project id (iirc that is hard to get) - thiugh I can prob set some up

parthea commented:
Getting all the tests working with a test project id could be done using the following steps:

[ ] Create a google account
[ ] Register for BigQuery, and create a ProjectId
[ ] Set the project id in `test_gbq.py`. Run the unit test and go through the authentication process.
[ ] Copy the file that is generated : `pandas/io/tests/bigquery_credentials.dat` 
[ ] Place the `bigquery_credentials.dat` into the `pandas/io/tests/` at the start of each build.

I have seen an advertisement from BigQuery with a 2 month(or $300) free trial.
https://cloud.google.com/free-trial/

jacobschaer commented:
We've tried several times to work something out with Google for this and haven't gotten anywhere. Ideally they would provide an account just for integration testing for this project - perhaps liming access to Travis IP, etc. @azbones - have we heard anything back from them?

@parthea - that's basically what we did.
1) Created BigQuery account
2) Ran gbq once to generate the credentials file
3) Put the credentials file on a server with Jenkins
4) Jenkins runs on commit, moving the credentials file to the correct place and doing a string substitution with the project ID (in the test files).

The catch was that Jenkins needed to run at least often enough to keep the token valid, but that turned out to not be too big of an issue.

parthea commented:
@jacobschaer
That's great! So everything is in place except a free account for integration testing which may take some time. Is it possible to use a BigQuery trial in the meantime? That should buy us 2 months. At the same time, we can understand what the cost would be for a paid account (you can see how long the $300 lasts in the trial), and it will also give us a change to tell Google what we're using it for in hopes that we can get an account for integration testing. It may help open the communications with Google.

jreback commented on ci/requirements-2.7_LOCALE.pip:
```diff
@@ -1 +1,3 @@
 blosc
+httplib2
+google-api-python-client == 1.2
```
you don't need to have this work in EVERY version. In fact let's just care about 2.7 (where you specify exact versions, e.g. the 1.2 client) and 3.4 (where you don't, e.g. the latest)

jreback commented:
@parthea1 let's open a new issue about proper testing of this on Travis. I am working on getting some credentials.

jreback commented:
@parthea1 update the whatsnew to show closing #10656 as well

parthea commented:
This should close #10652

jreback commented:
ok that is the correct issur

parthea committed:
BLD: Install google-api-python-client and httplib2 using pip

parthea committed:
BUG: Use StrictVersion instead of LooseVersion when testing for minimum google api client version #10652

azbones commented:
@jacobschaer I tried reaching out to Google's developer relations team and the BigQuery engineering manager several times, but they never came up with a solution for how to conduct testing with an account that would be public. It is unfortunate as I gave them a few different ideas about how to approach it (delete the data sets periodically, limit total data set size, etc.), but got no response.  You would think they would want to encourage this kind of work...

jreback commented:
@parthea1 ok, lgtm. ping on green.!

parthea commented:
Thanks! 

Travis is green. The only message in the skipped tests for gbq module is `Cannot run integration tests without a project id`

jreback commented:
@parthea1 big thanks! this was a really nice fixup!

jreback commented:
for your next trick...

https://github.com/blaze/odo/issues/285

@cpcloud manages this project and would be straightforward to add the routines needed to support BQ (a further and more complicated project would be to actually use a Blaze expression as the query, but as a first step io to/from are quite useful)

parthea commented:
I would definitely be up for working on the odo project as I enjoy a great challenge. I have a few work/personal commitments in the next few weeks. I will begin looking into the odo project in early October.
